course-mw/enrolment-actor/src/main/scala/org/sunbird/enrolments/ExtendedContentConsumptionActor.scala

```diff
@@ -0,0 +1,647 @@
+    val updateMap = new java.util.HashMap[String, AnyRef] () {{
+      put("lastreadcontentid", lastAccessContent.get(JsonKey.CONTENT_ID_KEY))
+      put("lastreadcontentstatus", lastAccessContent.get("status"))
+      put("lrc_progressdetails", lastAccessContent.get("progressdetails"))
+      put(JsonKey.LAST_CONTENT_ACCESS_TIME, lastAccessContent.get(JsonKey.LAST_ACCESS_TIME_KEY))
+
+    }}
+    val selectMap = new util.HashMap[String, AnyRef]() {{
+      put("batchId", batchId)
+      put("userId", userId)
+      put("courseId", lastAccessContent.get(JsonKey.COURSE_ID_KEY))
+    }}
+    (selectMap, updateMap)
+  }
+
+  @throws[Exception]
+  private def pushInstructionEvent(requestContext: RequestContext, userId: String, batchId: String, courseId: String, contents: java.util.List[java.util.Map[String, AnyRef]], primaryCategory:String, parentCollections: java.util.List[String]): Unit = {
+    val data = new java.util.HashMap[String, AnyRef]
+    data.put(CourseJsonKey.ACTOR, new java.util.HashMap[String, AnyRef]() {{
+      put(JsonKey.ID, InstructionEvent.BATCH_USER_STATE_UPDATE.getActorId)
+      put(JsonKey.TYPE, InstructionEvent.BATCH_USER_STATE_UPDATE.getActorType)
+    }})
+    data.put(CourseJsonKey.OBJECT, new java.util.HashMap[String, AnyRef]() {{
+      put(JsonKey.ID, batchId + CourseJsonKey.UNDERSCORE + userId)
+      put(JsonKey.TYPE, InstructionEvent.BATCH_USER_STATE_UPDATE.getType)
+    }})
+    data.put(CourseJsonKey.ACTION, InstructionEvent.BATCH_USER_STATE_UPDATE.getAction)
+    val contentsMap = contents.map(c => new java.util.HashMap[String, AnyRef]() {{
+      put(JsonKey.CONTENT_ID, c.get(JsonKey.CONTENT_ID_KEY))
+      put(JsonKey.STATUS, c.get(JsonKey.STATUS))
+    }}).asJava
+    data.put(CourseJsonKey.E_DATA, new java.util.HashMap[String, AnyRef]() {{
+      put(JsonKey.USER_ID, userId)
+      put(JsonKey.BATCH_ID, batchId)
+      put(JsonKey.COURSE_ID, courseId)
+      put(JsonKey.CONTENTS, contentsMap)
+      put(JsonKey.PRIMARYCATEGORY, primaryCategory)
+      put(JsonKey.PARENT_COLLECTIONS, parentCollections)
+      put(CourseJsonKey.ACTION, InstructionEvent.BATCH_USER_STATE_UPDATE.getAction)
+      put(CourseJsonKey.ITERATION, 1.asInstanceOf[AnyRef])
+    }})
+    val topic = ProjectUtil.getConfigValue("kafka_topics_instruction")
+    logger.info(requestContext,"LearnerStateUpdateActor: pushInstructionEvent :Event Data " + data + " and Topic " + topic)
+    if(pushTokafkaEnabled)
+      InstructionEventGenerator.pushInstructionEvent(userId, topic, data)
+  }
+
+  def getConsumption(request: Request): Unit = {
+    val userId = request.get(JsonKey.USER_ID).asInstanceOf[String]
+    val batchId = request.get(JsonKey.BATCH_ID).asInstanceOf[String]
+    val courseId = request.get(JsonKey.COURSE_ID).asInstanceOf[String]
+    val language = request.getOrDefault(JsonKey.LANGUAGE, "").asInstanceOf[String]
```

Comment:
why getOrDefault -- if this becomes empty - what would be the impact ??
before we start testing in dev with this code -- existing records must be migrated with the language details... so we must get the language details by now?
use object.get only -- let's check will there be any exception